docs/design-v2.md

```diff
@@ -0,0 +1,133 @@
+| ------------------------------------------- | ----------------------------------------- | ------------------------------------------------------ | --------- | ------------------------------------------------------------------------------------------------- |
+| `maxMountReplicaCount`                      | The number of mount replicas to maintain. | This value must be in the range `[0..(maxShares - 1)]` | No        | If `accessMode` is `ReadWriteMany`, the default is `0`. Otherwise, the default is `maxShares - 1` |
+| `enableFileSystemConsistencyCheck`          | ???                                       | `true` or `false`                                      | No        | `false`                                                                                           |
+| `enableMountReplicaReductionOnDiskPressure` | ???                                       | `true` or `false`                                      | No        | `false`                                                                                           |
+
+### Custom Resources and Controllers
+
+Mandalay uses 3 different custom resources defined in the `azure-disk-csi` namespace to 
+orchestrate disk management and facilitate pod failover to nodes with mount replicas. A 
+controller for each resource watches for and responds to changes to its custom resource 
+instances.
+
+#### `AzDriverNode` Resource and Controller
+
+The `AzDriverNode` custom resource represents a node in the cluster where the Mandalay node 
+plug-in runs. An instance of `AzDriverNode` is created when the node plug-in starts. The node 
+plug-in periodically updates the heartbeat in the `Status` field.
+
+The controller for `AzDriverNode` runs in the controller plug-in. It is responsible for 
+deleting `AzDriverNode` instances that no longer have corresponding nodes in the cluster.
+
+#### `AzVolume` Resource and Controller
+
+The `AzVolume` custom resource represents the managed disk of a `PersistentVolume`. The 
+controller for `AzVolume` runs in the controller plug-in. It watches for and reconciles 
+changes in the `AzVolume` instances.
+
+An `AzVolume` instance is created by the `CreateVolume` API to the CSI Controller plug-in. The 
+`AzVolume` controller responds to the new instance by creating a managed disk using the 
+parameters in the referenced `StorageClass`. When disk creation is complete, the controller 
+sets the `.Status.State` field to `Created` or suitable error state on failure. It also adds a 
+finalizer if creation was successful. The `CreateVolume` request completes when the `AzVolume` 
+state is updated.
+
+To delete a managed disk, the `DeleteVolume` API in the CSI Controller plug-in deletes the 
+corresponding `AzVolume` instance. The controller responds to deletion by garbage collecting 
+the managed disk based on the `reclaimPolicy` set in the `StorageClass`. When the reclaim 
```

Comment:
Isn't the handling for 'reclaimPolicy' already done by kubernetes controllers or does AzVolume controller decides to delete the managed disk based on the reclaimPolicy? Is DeleteVolume even called if the reclaimPolciy is 'retain'?

Reply:
Ah, yes, good point. I was also under the same impression, yet simply restated this from the original design doc. 

@abhisheksinghbaghel Does the `DeleteVolume` event get called when `reclaimPolicy` is set to `RETAIN`? Or was the intention here that garbage collection was more an asynchronous activity rather than done in response to the `AzVolume` CRI deletion?

Reply:
Yeah, the intention was to say that the actual disk deletion will happen async on AzVolume deletion.

For Retain reclaim policy. based on below, it doesn't look like we will get called. Which is very interesting:
https://github.com/kubernetes/kubernetes/blob/f6152d152140f42815765c7d9703b9817943c345/pkg/controller/volume/persistentvolume/pv_controller.go#L1082

It appears we will be leaking AzVolume objects, but looks like PV is also not deleted, so we should be good.

Reply:
@abhisheksinghbaghel I updated the description to drop the `reclaimPolicy` processing. I am assuming the `DeleteVolume` should not complete until the managed disk has been deleted. The CSI spec is clear on expectations here (other than requiring idempotency). Is this your expectation as well?

Reply:
@edreed I am not sure if I understand the second part of your comment. Our current plan is to delete the AzVolume object, and then controller lazily deletes the backing disk (we should check the reclaim policy here). Are you saying that we might have an issue due to that?

Reply:
I meant to say the "CSI spec is *not* clear on expectations here..." There may be a problem if the disk is deleted lazily. Since `CreateVolume` is idempotent, a subsequent call to create the PV with the same name (and characterstics) will succeed and pick up the ungarbage collected disk. This is an information disclosure threat.

Reply:
@edreed That's a good point. For all these cases, I think we should match the UIDs along with the names to make sure.
As for this specific example, we do mark the AzVolume deleted (set the delete timestamp) before responding back on the CSI call. So we should be able to do a validation on the incoming request and make sure that the disk is actually a different one and not in middle of deletion.

Reply:
@abhisheksinghbaghel I don't believe we capture the UID in order to make that check - I will investigate and confirm. Even if we rely on the AzVolume deletion state, we still wouldn't be able to create a new disk with the same name, could we?

Reply:
@edreed Yeah, disk names are unique on DRP (under a RG), so trying to recreate will result in a conflict.
Although, current handling for such cases is to say that the disk already exists, which could be a problem for us if delete and then create came too close (may be we can handle this corner case with a retryable error, because controller will delete the disk eventually?).

I think we are going to be ok for dynamic provisioning cases, because CO passes a unique Name in the PV request. We should see if we can run in to this problem with static provisioning (we should not be doing create here at all), or cases where we have hardcoded the PV name in the storageclass.